### PR TITLE
fix(snapshot): Refactor the big_value_mu_ mutex

### PR DESCRIPTION
**Pull request by BagritsevichStepan**
fixes dragonflydb#4152

### CODE REVIEW

File: src/server/snapshot.cc [hunk 1]
```diff
@@ -339,53 +348,59 @@ size_t SliceSnapshot::FlushSerialized(SerializerBase::FlushState flush_state) {
...
   DVLOG(2) << "Pushing " << id;
 
-  fb2::NoOpLock lk;
-
-  // We create a critical section here that ensures that records are pushed in sequential order.
-  // As a result, it is not possible for two fiber producers to push concurrently.
-  // If A.id = 5, and then B.id = 6, and both are blocked here, it means that last_pushed_id_ < 4.
-  // Once last_pushed_id_ = 4, A will be unblocked, while B will wait until A finishes pushing and
-  // update last_pushed_id_ to 5.
-  seq_cond_.wait(lk, [&] { return id == this->last_pushed_id_ + 1; });
```
Comment (BagritsevichStepan) [inline src/server/snapshot.cc:349]: I think we don't need this because FlushSerialized is called when big_value_mu_ is blocked

Comment (adiholden) [reply src/server/snapshot.cc:349]: maybe you are right here, but this is not the change I wanted by this PR.
I wanted the mutex to protect the serializer and only it.
So we will have a getter for the serializer when trying to access it and once done we release the mutex.
So in this point the mutex will not be taken

File: src/server/snapshot.cc [hunk 2]
```diff
@@ -318,18 +319,26 @@ void SliceSnapshot::SerializeEntry(DbIndex db_indx, const PrimeKey& pk, const Pr
...
+    });
     CHECK(res);
     ++type_freq_map_[*res];
   }
 }
 
 size_t SliceSnapshot::FlushSerialized(SerializerBase::FlushState flush_state) {
   io::StringFile sfile;
+
+  // FlushSerialized is already under lock
```
Comment (adiholden) [inline src/server/snapshot.cc:340]: The changes is not good
This is exactly what I wanted to simplify. I dont want us to think or review different flows to understand if the lock is already taken but instead have the getter protect us.

Comment (adiholden) [reply src/server/snapshot.cc:340]: I suggest revert now all this changes that are for the CallSerializerUnderLock
we can keep the other small changes that you did that does not impact the mutex.

Comment (BagritsevichStepan) [reply src/server/snapshot.cc:340]: The issue is that `FlushSerialized` can be invoked from the `flush_fn` callback passed to the `RdbSerializer`, which can be called during the `serializer->SaveEntry` method (where the mutex is already locked).

Comment (adiholden) [reply src/server/snapshot.cc:340]: Ok therefore I suggest reverting the changes as they do not simplify the code.
I will revisit this later and see if we can improve this

**Pull request closed without merging**